### DISCUSSION TITLE
Download progress in process

### DISCUSSION THREAD

**Discussion by Guitronic** [microsoft/winget-cli#5075] [General]
Hello all.
I am using winget in a Qt process and I can't get the download progress like with "powershell" for example. 
- Could someone tell me why I get the whole progress frame only at the end of the complete download?

**Comment by MusiqueGraeme** [comment]
I've written a test program that seems to do this... 
Func TestingProgramUpdate()
	$sProgTitle = "rclone"
	MsgBox(0,"",WingetAction("download " & $sProgTitle & " -d '" & @ScriptDir & "'" & (($sProgTitle = "VeraCrypt")?" --accept-package-agreements":"") & " --verbose-logs"))
EndFunc ;=>TestingProgramUpdate

WingetAction does most of the work. It starts a powershell command to download with Winget verbosely.
FindingOutput finds the file name of the verbose log
FindDownloadSpeed uses a winmgmnts query to get total downloaded bits which it takes from the number when the program started to compare this with the total of the file to be downloaded.

Func WingetAction($Parameters)
	$sStartTime = @YEAR & @MON & @MDAY & @HOUR & @MIN & @SEC
	SplashTextOn($PrettyScriptName,"Looking for Download",200,50,@DesktopWidth -270,@DesktopHeight / 2, 16)
	$nTaskOutput = Run(@ComSpec & " /c " & 'powershell winget ' & $Parameters,'',@SW_HIDE,$STDOUT_CHILD)
	Sleep(3800)
	$sLogFile = FindingOutput($sStartTime)
	If StringRight($sLogFile,1) = "\" Then
		Sleep(1000)
		$sLogFile = FindingOutput($sStartTime)
	EndIf
	$sTempData = ""
	$sTempPath = ""
	$nTempSize = 0
	While ProcessExists($nTaskOutput)
		$sLogData = FileRead($sLogFile)
		$sOutData = ($sTempData <> "")?StringReplace($sLogData, $sTempData,""):$sLogData
		If StringInStr($sOutData,"Found subcommand: download") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 10%",1)
		If StringInStr($sOutData,"Found matching extension.") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 12%",1)
		If StringInStr($sOutData,"Starting installer selection.") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 15%",1)
		If StringInStr($sOutData,"Completed installer selection.") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 18%",1)
		If StringInStr($sOutData,"Downloading to path:") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 20%",1)
		If StringInStr($sOutData,"DeliveryOptimization downloading from url:") > 0	Then
			$nStart = StringInStr($sOutData,"DeliveryOptimization downloading from url:") + StringLen("DeliveryOptimization downloading from url:")
			$nEndText = StringInStr($sOutData,@CRLF,0,1,$nStart)
			$sTempPath = StringMid($sOutData,$nStart,$nEndText - $nStart)
			$nTempSize = InetGetSize($sTempPath)
			ConsoleWrite("Downloading file from " & $sTempPath & " of " & $nTempSize)
			$nBaseDownload = FindDownloadedBits()
		EndIf
		If $nTempSize > 0 Then
			$nDownloadGuess = FindDownloadedBits() - $nBaseDownload
			$nPercentDone = 20 + Round($nDownloadGuess/$nTempSize * 50,0)
			ControlSetText($PrettyScriptName,"","Static1","Download " & (($nPercentDone < 70)?$nPercentDone:70) & "%",1)
		EndIf
		If StringInStr($sOutData,"Download completed.") > 0	Then
			ControlSetText($PrettyScriptName,"","Static1","Download 70%",1)
			$nTempSize = 0
		EndIf
		If StringInStr($sOutData,"Started applying motw to") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 80%",1)
		If StringInStr($sOutData,"Installer hash verified") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 90%",1)
		If StringInStr($sOutData,"Leaf command succeeded: root:download") > 0	Then ControlSetText($PrettyScriptName,"","Static1","Download 100%",1)
		$sTempData = $sLogData
		Sleep(150)
	WEnd
	$sOutData = StdoutRead($nTaskOutput)
	ConsoleWrite($sOutData)
	Return StringMid($sOutData,StringInStr($sOutData,"Installer downloaded: ") + StringLen("Installer downloaded: "))
EndFunc ;=>WingetAction

Func FindingOutput($sStart)
	$sWingetLogFile = ""
	$nCount = 1
	$sWingetLogPath = StringReplace(@AppDataDir,"Roaming","Local\Packages\")
	$hFile = FileFindFirstFile($sWingetLogPath & "Microsoft.Desktop*.*")
	$sWingetLogPath &= FileFindNextFile($hFile) & "\LocalState\DiagOutputDir\"
	$hFile = FileFindFirstFile($sWingetLogPath & "*.*")
	While 1
		$sTempFile = FileFindNextFile($hFile)
		If @error = 1 Then ExitLoop
		If FileGetTime($sWingetLogPath & $sTempFile,1,1) > $sStart Then $sWingetLogFile = $sTempFile
	WEnd
	ConsoleWrite($sWingetLogPath & $sWingetLogFile & @CRLF)
	Return $sWingetLogPath & $sWingetLogFile
EndFunc ;=>FindingOutput

Func FindDownloadedBits()

$objWMIService = ObjGet("winmgmts:\\" & @ComputerName & "\root\CIMV2")
Dim $colItems = $objWMIService.ExecQuery("SELECT * FROM Win32_PerfRawData_Tcpip_NetworkInterface")
$nDownloadedBits = 0
For $objItem In $colItems
    $nDownloadedBits += $objItem.BytesReceivedPersec
   ; $uploadSpeed = $objItem.BytesSentPersec
    ConsoleWrite("Download: " & $nDownloadedBits & " B/s " & @CRLF)
Next
Return $nDownloadedBits
EndFunc ;=>FindDownloadedBits
